// webhookrelay-ws-client/lib/Messages/WebhookRelayEvent.js
"use strict";
// {
//     "type": "webhook",             // event type
//     "meta": {                      // bucket, input and output information 
//       "bucked_id": "1593fe5f-45f9-45cc-ba23-675fdc7c1638", 
//       "bucket_name": "my-1-bucket-name",                                
//       "input_id": "b90f2fe9-621d-4290-9e74-edd5b61325dd",
//       "input_name": "Default public endpoint",
//       "output_name": "111",
//       "output_destination": "http://localhost:8080"
//     },
//     "headers": {                   // request headers
//       "Content-Type": [
//         "application/json"
//       ]
//     },
//     "query": "foo=bar",            // query (ie: /some-path?foo=bar)
//     "body": "{\"hi\": \"there\"}", // request body
//     "method": "PUT"                // request method
//   }
Object.defineProperty(exports, "__esModule", { value: true });
var SubscriptionMessage = /** @class */ (function () {
    /** @private */
    // constructor(readonly _data: Event) {}
    function SubscriptionMessage(type, status, message, meta, headers, query, body, method) {
        this.type = type;
        this.status = status;
        this.message = message;
        this.meta = meta;
        this.headers = headers;
        this.query = query;
        this.body = body;
        this.method = method;
    }
    /**
     * Name of the bucket that received the webhook
     */
    SubscriptionMessage.prototype.getBucketName = function () {
        return this.meta.bucket_id;
    };
    SubscriptionMessage.prototype.getType = function () {
        return this.type;
    };
    SubscriptionMessage.prototype.getStatus = function () {
        return this.status;
    };
    SubscriptionMessage.prototype.getMessage = function () {
        return this.message;
    };
    SubscriptionMessage.fromJSON = function (json) {
        var msg = Object.create(SubscriptionMessage.prototype);
        return Object.assign(msg, json, {
            created: new Date(json.created)
        });
    };
    return SubscriptionMessage;
}());
exports.SubscriptionMessage = SubscriptionMessage;
exports.default = SubscriptionMessage;
var ResponseMessage = /** @class */ (function () {
    /** @private */
    // constructor(readonly _data: ResponseMessage) {}
    function ResponseMessage(meta, status, headers, body) {
        this.meta = meta;
        this.status = status;
        this.headers = headers;
        this.body = body;
    }
    ResponseMessage.prototype.getMeta = function () {
        return this.meta;
    };
    ResponseMessage.prototype.getStatus = function () {
        return this.status;
    };
    ResponseMessage.prototype.getHeaders = function () {
        return this.headers;
    };
    ResponseMessage.prototype.getBody = function () {
        return this.body;
    };
    ResponseMessage.fromJSON = function (json) {
        var msg = Object.create(ResponseMessage.prototype);
        return Object.assign(msg, json, {
            created: new Date(json.created)
        });
    };
    return ResponseMessage;
}());
exports.ResponseMessage = ResponseMessage;
